woodoku/entity/score_agent.py

```diff
@@ -26,19 +28,30 @@ def calculate_winning(self, blocks: int, groups: int) -> None:
             blocks (int): The number of blocks placed
             groups (int): The number of groups completed
         """
-        self.__score += blocks
+        # Note: assertion can be turned off with -O flag to python
+        assert blocks > 0 and groups >= 0, "Precondition violated"
+
+        # Note: instead of self.__score += blocks. Do
+        self.__score = self.get_score() + blocks
+        # this way get_score() can be mocked for easier testing setup ;)
 
         if groups:
-            self.__score += GROUP_POINTS
-            if self.__streaks:
-                self.__score += STREAK_POINTS * self.__streaks
-            self.__streaks += 1
+            self.__score = self.get_score() + GROUP_POINTS
+            if self.__streak:
+                self.__score = self.get_score() + STREAK_POINTS * self.get_streak()
+            self.__streak = self.get_streak() + 1
         else:
-            self.__streaks = 0
+            self.__streak = 0
 
-        combos = groups - 1
-        if combos > 0:
-            self.__score += COMBO_POINTS * combos
+        self.__combo = groups - 1
+        if self.get_combo() > 0:
+            self.__score = self.get_score() + COMBO_POINTS * self.get_combo()
```

Comment:
Why calling method instead of accessing the instance variables?

Reply:
This is for mocking purposes. Since get is public methods, I can use them in combination with calculating winning to more easily set the private state of a score agent. Otherwise, it would be the current implementation that reserves engineer the desired score and how to set the state only with calculate winning(which can't be mocked since it's a void method)

Reply:
I'd need these get method to implement the mocking in #61 so I'd probably will just copy/paste these method or cherry pick the commit of these changes to #61methods

Reply:
reversing course on mocking see comment[https://github.com/EdwardHaoranLee/WoodokuLearn/pull/61#discussion_r936237242](url) here, so I will reverse and access instance variables directly again